src/clj/puppetlabs/services/jruby/jruby_internal.clj

```diff
@@ -139,39 +146,42 @@
 (schema/defn ^:always-validate
   cleanup-pool-instance!
   "Cleans up and cleanly terminates a JRubyInstance and removes it from the pool."
-  [{:keys [scripting-container pool] :as instance} :- JRubyInstance]
+  [{:keys [scripting-container pool] :as instance} :- jruby-schemas/JRubyInstanceSchema
+   shutdown-fn :- IFn]
   (.unregister pool instance)
-  ;; TODO: need to add support for a callback hook, so that consumers like
-  ;; puppet-server can do their own cleanup.
-  ;(.terminate jruby-puppet)
+  (shutdown-fn instance)
   (.terminate scripting-container)
   (log/infof "Cleaned up old JRubyInstance with id %s." (:id instance)))
 
 (schema/defn ^:always-validate
-  create-pool-instance! :- JRubyInstance
+  create-pool-instance! :- jruby-schemas/JRubyInstanceSchema
   "Creates a new JRubyInstance and adds it to the pool."
   [pool :- jruby-schemas/pool-queue-type
    id :- schema/Int
    config :- jruby-schemas/JRubyConfig
    flush-instance-fn :- IFn]
-  (let [{:keys [ruby-load-path gem-home compile-mode]} config]
+  (let [{:keys [ruby-load-path gem-home compile-mode lifecycle]} config
+        init-fn (:initialize lifecycle)
+        init-env-vars-fn (:initialize-env-variables lifecycle)]
```

Comment:
Yeah, I'm trying to remember what arguments I made for trying to avoid this call...

I think I was concerned about allowing the consumer to be able to have access to the full `ScriptingContainer` instance so that the consumer could coordinate whatever current values are available in it with what gets used for the environment variables.  The design on this PR appears to allow for this still since the full `ScriptingContainer` instance appears to be passed in on the callback.

The other thing I hadn't mentioned at the time - but seems important - is that the consumer might want to just have access to the full set of default environment variables as JRuby itself would first set them up on a container instance rather than only get a chance to reset them later after our default logic had filtered them out.  For this case, I also think that what is on this PR would be better than what I'd suggested previously.

Where the kinds of actions that the consumer can perform during this lifecycle function wouldn't technically be limited to just setting up environment variables, I wonder, though, if it might be better to name this lifecycle function differently.  Like maybe we could have:

-  an `initialize-scripting-container` lifecycle function which is for performing any changes to the `ScriptingContainer` before the jruby-utils code would do any setup that involves calling `.runScriptlet` (after which point any further changes may be ignored).

... and ...

- an `initialize-pool-instance` lifecycle function which is about producing the instance that goes into the pool (but could receive a `ScriptingContainer` instance that had already had some Ruby code executed within it).